### PR TITLE
SAA-1057 introduction of new entity support prisoner waitlist functionality.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/WaitingList.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  val prisonerNumber: String,
+
+  val bookingId: Long,
+
+  val applicationDate: LocalDate,
+
+  @OneToOne
+  @JoinColumn(name = "activity_schedule_id", nullable = false)
+  val activitySchedule: ActivitySchedule,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/WaitingList.kt:32]: @timharrison-moj I struggle not to model this as I would like to see it.

Totally aware of the instances issue.  Would you be happy if we were to limit the instances returned through the use of repo filters?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/WaitingList.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  @JoinColumn(name = "allocation_id", nullable = false)
+  var allocation: Allocation? = null
+}
+
+enum class WaitingListStatus {
+  PENDING,
+  APPROVED,
+  DECLINED,
+  // TODO add more statuses
+}
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/WaitingList.kt:59]: 2 more likely here - ALLOCATED (approved and allocated) , REMOVED (allocated and removed).

File: src/main/resources/migrations/common/V2023.07.31__add_waiting_list.sql [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  creation_time        timestamp    NOT NULL,
+  created_by           varchar(20)  NOT NULL,
+  comments             varchar(300),
+  declined_reason      varchar(100),
+  updated_time         timestamp,
+  updated_by           varchar(100),
+  allocation_id        bigint REFERENCES allocation (allocation_id)
+);
+
+CREATE INDEX idx_waiting_list_allocation_id ON waiting_list (allocation_id);
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply src/main/resources/migrations/common/V2023.07.31__add_waiting_list.sql:20]: From memory, I think waiting lists are quite long - there are more people waiting for activities than are allocated - that kind of size, so indexes on prison_code, prisoner_number, application_date.   

The allocation_id should not reference allocation though - and should be nullable - waiting and approved rows do not have am allocation id.

[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/resources/migrations/common/V2023.07.31__add_waiting_list.sql:20]: Duh of course. Facepalm.

[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/resources/migrations/common/V2023.07.31__add_waiting_list.sql]: Added indexes.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/WaitingList.kt [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+  var declinedReason: String? = null
+
+  var updatedTime: LocalDateTime? = null
+
+  var updatedBy: String? = null
+
+  @OneToOne
+  @JoinColumn(name = "allocation_id", nullable = false)
+  var allocation: Allocation? = null
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/WaitingList.kt:51]: nullable = true?

**Review by timharrison-moj — DISMISSED** [review]
LGTM. Quick check on nullable = false, and whether that only applies if a value is present?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/WaitingList.kt (hunk 4 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/WaitingList.kt:51]: Doh

Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/WaitingList.kt:51]: done

**Review by timharrison-moj — DISMISSED** [review]
LGTM

File: src/main/resources/migrations/common/V2023.08.01__add_waiting_list.sql [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  prisoner_number      varchar(7)   NOT NULL,
+  booking_id           bigint       NOT NULL,
+  application_date     date         NOT NULL,
+  activity_id          bigint       NOT NULL,
+  activity_schedule_id bigint       NOT NULL,
+  requested_by         varchar(100) NOT NULL,
+  status               varchar(20)  NOT NULL,
+  creation_time        timestamp    NOT NULL,
+  created_by           varchar(20)  NOT NULL,
+  comments             varchar(500),
```
Comment (matt-crowson-moj) [inline src/main/resources/migrations/common/V2023.08.01__add_waiting_list.sql:13]: @timharrison-moj can you reapprove please.  Turns out the JIRA ticket specified 500 characters for the comments field.

**Review by timharrison-moj — APPROVED** [review]
LGTM.

**Review by timharrison-moj — APPROVED** [review]
LGTM

**Pull request merged**